logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java

```diff
@@ -67,7 +79,25 @@ public void write(String fileName, Checkpoint checkpoint) throws IOException {
             out.getChannel().write(buffer);
             out.getFD().sync();
         }
-        Files.move(tmpPath, dirPath.resolve(fileName), StandardCopyOption.ATOMIC_MOVE);
+
+        try {
+            Files.move(tmpPath, dirPath.resolve(fileName), StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException ex) {
+            if (retry) {
+                try {
+                    logger.error("Retrying after exception writing checkpoint: " + ex);
+                    Thread.sleep(500);
+                    Files.move(tmpPath, dirPath.resolve(fileName), StandardCopyOption.ATOMIC_MOVE);
+                } catch (Exception ex2) {
+                    logger.error("Aborting after second exception writing checkpoint: " + ex2);
+                    throw new IllegalStateException(ex2);
```

Comment:
any reason to not just catch IOException and rethrow instead of producing an IllegalStateException?
Regardless of having retried once or not, failing to write the checkpoint should have a single outcome.